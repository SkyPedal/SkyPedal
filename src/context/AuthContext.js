import { createContext, useContext } from "react";

const AuthContext = createContext();

export default AuthContext;

export const useAuth = () => {
  return useContext(AuthContext);
};
