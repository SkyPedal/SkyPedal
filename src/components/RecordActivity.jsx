import { useEffect, useState } from "react";
import { useAuth } from "../context/AuthContext";
import ActivityForm from "./record/ActivityForm";
import today from "../utils/today";
import { useNavigate } from "react-router-dom";
import { getLocations, saveActivity, getRoute } from "../repos/api";

const RecordActivity = () => {
  const [locations, setLocations] = useState([]);

  const [error, setError] = useState("");
  const [formData, setFormData] = useState({
    title: "Commute to Work",
    date: today(),
    time: "08:00",
    start: "none",
    end: "none",
    distance: 0,
    duration: 0,
  });
  const { user_id } = useAuth();

  const navigate = useNavigate();

  useEffect(() => {
    getLocations(user_id).then((data) => {
      if (data.error) {
        setError(data.error);
      } else {
        setLocations(data.data);
      }
    });
  }, [user_id]);

  const handleSave = async (e) => {
    e.preventDefault();
    const activity = {
      ...formData,
    };
    setError("");
    const success = await saveActivity(activity, user_id);
    console.log("Submit: ", success, error);
    navigate("/");
  };

  useEffect(() => {
    if (
      formData.start !== "Add New" &&
      formData.start !== "none" &&
      formData.end !== "Add New" &&
      formData.end !== "none"
    ) {
      setError("");
      getRoute(formData.start, formData.end).then((data) => {
        if (data.error) {
          setError(data.error);
          return;
        }
        setFormData((prevData) => ({
          ...prevData,
          distance: data.data.distance,
          duration: data.data.duration,
        }));
      });
    }
  }, [formData.start, formData.end]);

  return (
    <div className="relative mx-auto h-full max-w-xl p-4">
      <h1 className="pt-5 text-left text-3xl">Record Activity</h1>
      {error != "" && <span className="text-red-500">{error}</span>}
      <ActivityForm
        handleSave={handleSave}
        locations={locations}
        formData={formData}
        setFormData={setFormData}
      />
    </div>
  );
};

export default RecordActivity;
