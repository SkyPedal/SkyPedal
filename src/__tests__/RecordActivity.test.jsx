import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import RecordActivity from "../components/RecordActivity";
import { BrowserRouter as Router } from "react-router-dom";
import AuthContext from "../context/AuthContext";

describe("RecordActivity", () => {
  it("snapshot renders", () => {
    const { asFragment } = render(
      <Router>
        <AuthContext.Provider value={{ user_id: 1, user_name: "TestUser123" }}>
          <RecordActivity />
        </AuthContext.Provider>
      </Router>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
