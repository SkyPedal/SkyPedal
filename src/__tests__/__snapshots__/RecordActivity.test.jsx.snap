// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RecordActivity > snapshot renders 1`] = `
<DocumentFragment>
  <div
    class="relative mx-auto h-full max-w-xl p-4"
  >
    <h1
      class="pt-5 text-left text-3xl"
    >
      Record Activity
    </h1>
    <form>
      <div
        class="flex flex-col pt-12"
      >
        <label
          class="text-left text-lg"
          for="activityName"
        >
          Title
        </label>
        <input
          class="rounded-lg border-2 border-solid border-gray-300 p-2"
          id="activityName"
          name="activityName"
          type="text"
          value="Commute to Work"
        />
        <div
          class="flex flex-row justify-between"
        >
          <div
            class="flex flex-col"
          >
            <label
              class="pt-5 text-left text-lg"
              for="activityDate"
            >
              Date
            </label>
            <input
              class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
              id="activityDate"
              name="activityDate"
              type="date"
              value="2024-07-31"
            />
          </div>
          <div
            class="flex flex-col"
          >
            <label
              class="pt-5 text-left text-lg"
              for="activityTime"
            >
              Time
            </label>
            <input
              class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
              id="activityTime"
              name="activityTime"
              type="time"
              value="08:00"
            />
          </div>
        </div>
        <div
          class="mt-12"
        >
          <div
            class="flex flex-row justify-between"
          >
            <div
              class="flex flex-col"
            >
              <label
                class="pt-5 text-left text-lg"
                for="activityStart"
              >
                Start Location
              </label>
              <input
                class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
                disabled=""
                id="activityStart"
                name="activityStart"
                type="text"
                value=""
              />
            </div>
            <div
              class="flex flex-col"
            >
              <label
                class="pt-5 text-left text-lg"
                for="activityEnd"
              >
                End Location
              </label>
              <input
                class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
                disabled=""
                id="activityEnd"
                name="activityEnd"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <p
          class="mb-2 mt-8 w-full border-b-2 text-center"
          style="line-height: 0.1em;"
        >
          <span
            class="bg-white pl-2 pr-2"
          >
            OR
          </span>
        </p>
        <div
          class="flex flex-row justify-between"
        >
          <div
            class="flex flex-col"
          >
            <label
              class="pt-5 text-left text-lg"
              for="activityDistance"
            >
              Distance (miles)
            </label>
            <input
              class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
              id="activityDistance"
              min="1"
              name="activityDistance"
              required=""
              type="number"
              value="0"
            />
          </div>
          <div
            class="flex flex-col"
          >
            <label
              class="pt-5 text-left text-lg"
              for="activityDuration"
            >
              Duration (minutes)
            </label>
            <input
              class="w-52 rounded-lg border-2 border-solid border-gray-300 p-2"
              id="activityDuration"
              min="1"
              name="activityDuration"
              required=""
              type="number"
              value="0"
            />
          </div>
        </div>
      </div>
      <div
        class="absolute bottom-12 right-4 flex w-full flex-row justify-end"
      >
        <button
          class="mx-8 mt-5 max-h-10 w-24 rounded bg-red-500 px-4 py-2 font-bold text-white hover:bg-red-700"
          type="button"
        >
          Cancel
        </button>
        <button
          class="mt-5 max-h-10 w-24 rounded bg-blue-500 px-4 py-2 font-bold text-white hover:bg-blue-700"
          type="submit"
        >
          Add
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
